action/protocol/staking/candidate_center.go

```diff
@@ -7,143 +7,432 @@
 package staking
 
 import (
+	"sync"
+
 	"github.com/iotexproject/iotex-address/address"
 )
 
 type (
-	// CandidateCenter is a struct to manage the candidates
-	CandidateCenter struct {
+	// CandidateControl control the candidate center
+	CandidateControl interface {
+		CandidateView
+		Base() CandidateControl
+		Delta() CandidateList
+		SetDelta(CandidateList) error
+		Commit(CandidateList) error
+	}
+
+	// CandidateView provides access to all candidates
+	CandidateView interface {
+		Size() int
+		All() CandidateList
+		ContainsName(string) bool
+		ContainsOwner(address.Address) bool
+		ContainsOperator(address.Address) bool
+		ContainsSelfStakingBucket(uint64) bool
+		GetByName(string) *Candidate
+		GetByOwner(address.Address) *Candidate
+		GetBySelfStakingIndex(uint64) *Candidate
+		Upsert(*Candidate) error
+		collision(*Candidate) error
+	}
+
+	// candChange captures the change to candidates
+	candChange struct {
+		dirty map[string]*Candidate
+	}
+
+	// candBase is the confirmed base state
+	candBase struct {
+		lock             sync.RWMutex
 		nameMap          map[string]*Candidate
 		ownerMap         map[string]*Candidate
 		operatorMap      map[string]*Candidate
 		selfStkBucketMap map[uint64]*Candidate
 	}
+
+	// CandidateCenter is a struct to manage the candidates
+	CandidateCenter struct {
+		*candBase
+		size   int
+		change CandidateView
```

Comment:
if lock is on "candBase" level, it shouldn't be exposed to CandidateCenter or used in the CandidateCenter level.

Reply:
fixed in latest push